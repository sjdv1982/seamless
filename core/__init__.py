import logging

# Module logger
logger = logging.getLogger(__name__)
