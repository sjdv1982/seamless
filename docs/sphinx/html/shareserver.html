<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>The Seamless shareserver &mdash; Seamless 0.10 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=acd5942f" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5be48651"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=c71f9630"></script>
        <script src="_static/doctools.js?v=695de88e"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The help system" href="help.html" />
    <link rel="prev" title="Mounting to the file system" href="mounting.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Seamless
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="whatis.html">Seamless: a cell-based interactive workflow framework</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started with Seamless</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="explained.html">Seamless explained</a></li>
<li class="toctree-l1"><a class="reference internal" href="new-project.html">Seamless projects</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Seamless features</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="context.html">Contexts</a></li>
<li class="toctree-l1"><a class="reference internal" href="cell.html">Cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="transformer.html">Transformers</a></li>
<li class="toctree-l1"><a class="reference internal" href="mounting.html">Mounting to the file system</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">The Seamless shareserver</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">The help system</a></li>
<li class="toctree-l1"><a class="reference internal" href="visualization.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="validation.html">Validation</a></li>
<li class="toctree-l1"><a class="reference internal" href="libraries.html">Libraries and libinstances</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="debugging.html">Debugging</a></li>
<li class="toctree-l1"><a class="reference internal" href="environments.html">Environments</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_storage.html">Data storage</a></li>
<li class="toctree-l1"><a class="reference internal" href="job_management.html">Job management</a></li>
<li class="toctree-l1"><a class="reference internal" href="deployment.html">Deployment</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Advanced features</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="deepcell.html">Deep cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="low_level.html">Macros and the low level</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">stdlib documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="stdlib/merge.html">stdlib.merge</a></li>
<li class="toctree-l1"><a class="reference internal" href="stdlib/select.html">stdlib.select</a></li>
<li class="toctree-l1"><a class="reference internal" href="stdlib/switch.html">stdlib.switch</a></li>
<li class="toctree-l1"><a class="reference internal" href="stdlib/instantiate.html">stdlib.instantiate</a></li>
<li class="toctree-l1"><a class="reference internal" href="stdlib/map_dict.html">stdlib.map.map_dict</a></li>
<li class="toctree-l1"><a class="reference internal" href="stdlib/map_dict_chunk.html">stdlib.map.map_dict_chunk</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Seamless</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">The Seamless shareserver</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/shareserver.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="the-seamless-shareserver">
<h1>The Seamless shareserver<a class="headerlink" href="#the-seamless-shareserver" title="Link to this heading"></a></h1>
<p>The purpose of the shareserver is that you can share cell values over HTTP. That means that you can see the cell value when you type its URL in your browser. In addition, if you use a Seamless project, a web interface is automatically generated from all shared cells. By default, shared cells are readonly, and can only be visualized. Shared cells with readonly disabled can also be edited via a web interface.</p>
<p><em><strong>IMPORTANT: This documentation section is an early draft. The raw text material is shown below</strong></em></p>
<p><strong>Relevant examples:</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/sjdv1982/seamless/tree/stable/examples">basic example</a></p></li>
<li><p><a class="reference external" href="https://github.com/sjdv1982/seamless/tree/stable/examples/webserver-demo">webserver demo</a></p></li>
<li><p><a class="reference external" href="https://github.com/sjdv1982/seamless/tree/stable/examples/webserver-example">webserver example</a></p></li>
<li><p><a class="reference external" href="https://github.com/sjdv1982/seamless/tree/stable/examples/datatables-example">datatables</a></p></li>
<li><p><a class="reference external" href="https://github.com/sjdv1982/seamless/tree/stable/examples/grid-editor-example">grid editor</a></p></li>
</ul>
<p>The shareserver is to share Seamless cells via its REST API.
By default, the REST port is 5813
In addition, the shareserver sends notifications via Websocket connections, served at port 5138
The primary ways to interact with the share server are:</p>
<ul class="simple">
<li><p>The browser, for GET requests.
http://localhost:5813/ctx/a will get the value of cell “a”, if it has been shared</p></li>
<li><p>The seamless-http-put command, for PUT requests.</p></li>
<li><p>The seamless client (seamless-client.js)</p></li>
</ul>
<!--
## HTTP sharing

Intro:

- share
- cell.mimetype
- REST
- Your own web page, Seamless as file server (link to beginner gotchas)
--></section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="mounting.html" class="btn btn-neutral float-left" title="Mounting to the file system" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="help.html" class="btn btn-neutral float-right" title="The help system" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2016-2023, INSERM, CNRS and code contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>