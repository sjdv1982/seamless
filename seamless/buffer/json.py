import orjson

def json_encode(obj):
    dump = orjson.dumps(obj, option=orjson.OPT_SERIALIZE_NUMPY)
    return dump.decode()

def json_dumps(obj, as_bytes=False):
    dump = orjson.dumps(obj, option=orjson.OPT_INDENT_2 | orjson.OPT_SORT_KEYS)
    if not as_bytes:
        dump = dump.decode()
    return dump