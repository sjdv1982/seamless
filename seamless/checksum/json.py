"""Seamless's implementation to convert from/to JSON
This is the basis of the "plain" celltype."""

import orjson


def json_encode(obj) -> str:
    """Encode as JSON, tolerating Numpy objects"""
    dump = orjson.dumps(obj, option=orjson.OPT_SERIALIZE_NUMPY)
    return dump.decode()


def json_dumps(obj, as_bytes: bool = False) -> str | bytes:
    """Encode as JSON, with two-space identation and sorted keys"""
    dump = orjson.dumps(obj, option=orjson.OPT_INDENT_2 | orjson.OPT_SORT_KEYS)
    if not as_bytes:
        dump = dump.decode()
    return dump
