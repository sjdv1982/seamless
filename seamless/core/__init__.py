"""Deprecation stub"""

import warnings
import seamless.workflow.core
from seamless.workflow.core import *

warnings.warn(
    "seamless.core is deprecated. Use seamless.workflow.core instead.",
    DeprecationWarning,
    2,
)

__all__ = list(dir(seamless.workflow.core))
