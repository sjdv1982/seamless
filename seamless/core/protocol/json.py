'''
import json
import numpy as np
from silk.mixed import MixedScalar
from silk.validation import _integer_types, _float_types

def seamless_encoder(obj):
    retry = False
    if isinstance(obj, MixedScalar):
        obj = obj.value
        retry = True
    if isinstance(obj, np.generic):
        if isinstance(obj, _integer_types):
            obj = int(obj)
            retry = True
        elif isinstance(obj, _float_types):
            obj = float(obj)
            retry = True
    if retry:
        return default_encoder.encode(obj)
    else:
        typename = obj.__class__.__name__
        raise TypeError("%s object %s is not JSON serializable" % (typename, repr(obj)))

def json_encode(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
        allow_nan=True, cls=None, indent=None, separators=None,
        default=seamless_encoder, sort_keys=False, **kw):
    """Serialize ``obj`` to a JSON formatted ``str``.

    Wrapper around json.dumps that overrides the default value of "default"
    """
    global default_encoder
    default_encoder = json.JSONEncoder(
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      indent=indent,
      separators=separators,
      default=None,
      sort_keys=sort_keys,
      **kw
    )
    return json.dumps(obj,
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      cls=cls,
      indent=indent,
      separators=separators,
      default=default,
      sort_keys=sort_keys,
      **kw
    )

def json_dumps(obj, as_bytes=False):
    dump = json.dumps(obj, indent=2, sort_keys=True)
    if as_bytes:
        dump = dump.encode()
    return dump
'''

import orjson

def json_encode(obj):
    dump = orjson.dumps(obj, option=orjson.OPT_SERIALIZE_NUMPY)
    return dump.decode()

def json_dumps(obj, as_bytes=False):
    dump = orjson.dumps(obj, option=orjson.OPT_INDENT_2 | orjson.OPT_SORT_KEYS)
    if not as_bytes:
        dump = dump.decode()
    return dump