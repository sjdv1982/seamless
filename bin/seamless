#!/usr/bin/env -S python3 -u

__version__ = "0.12"

import argparse
from copy import copy
import sys
import os
from pathlib import Path
import json
import traceback
import seamless

from seamless.core.cache.buffer_cache import buffer_cache
from seamless.core.protocol.deserialize import deserialize_sync as deserialize
from seamless import CacheMissError
from seamless.cmd.message import set_verbosity, message as msg
from seamless.cmd import parsing
from seamless.cmd.parsing import get_commands, guess_arguments
from seamless.cmd.file_mapping import get_file_mapping
from seamless.cmd.file_load import files_to_checksums
from seamless.cmd.bash_transformation import prepare_bash_transformation, run_transformation
from seamless.cmd import interface
from seamless.highlevel import Checksum

parser = argparse.ArgumentParser()
parser.add_argument(
    "-v",
    dest="verbosity",
    help="""Verbose mode.
Multiple -v options increase the verbosity. The maximum is 3""",
    action="count",
    default=0,
)
parser.add_argument(
    "-q", dest="verbosity", help="Quiet mode", action="store_const", const=-1
)

parser.add_argument(
    "-g1",
    help="""Disable argtyping guess rule 1.
This rule states that any argument with a file extension must exist as a file.""",
    action="store_true",
)

parser.add_argument(
    "-g2",
    help="""Disable argtyping guess rule 2.
This rule states that any argument without file extension must not exist as a file.""",
    action="store_true",
)

parser.add_argument(
    "-w",
    help="""Set the argtyping working directory to DIR.
This allows any file in DIR to be specified as argument.""",
    dest="workdir",
)

parser.add_argument(
    "-W",
    help="""Set the argtyping working directory to /.
This allows any file to be specified as argument.""",
    dest="workdir",
    action="store_const",
    const="/",
)

parser.add_argument(
    "-ms",
    help="""Set the argtyping file mapping mode to 'strip'.
Strip directory names. After stripping, all files must be unique.""",
    dest="file_mapping_mode",
    action="store_const",
    const="strip",
)


parser.add_argument(
    "-mx",
    help="""Set the argtyping file mapping mode to 'extension'.
Keep only file extensions. Rename the file name body to file1, file2, ...
Example: 
  python script.py data.txt 
  => 
  python file1.py file2.txt""",
    dest="file_mapping_mode",
    action="store_const",
    const="extension",
)

parser.add_argument(
    "-c", help="Unquote the command line", dest="unquote", action="store_true"
)

parser.add_argument(
    "--undo",
    help="Undo this seamless command",
    action="store_true"
)

parser.add_argument('command', nargs=argparse.REMAINDER)

args = parser.parse_args()
command = args.command


verbosity = min(args.verbosity, 3)
set_verbosity(verbosity)
msg(1, "Verbosity set to {}".format(verbosity))
msg(1, "seamless {}".format(__version__))
msg(3, "Command:", json.dumps(command, indent=4))

if len(command) == 0:
    parser.print_usage()  # TODO: add to usage message
    sys.exit(0)

################################################################

if args.unquote:
    if len(command) != 1:
        msg(-1, "Unquote requires a single argument")
        sys.exit(1)
    commandstring = command[0]
else:
    commandstring = " ".join(command)

commands = get_commands(commandstring)

first_command = commands[0]

interface_argindex, interface_file, interface_py_file, mapped_execarg = interface.locate_files(first_command.words)

msg(1, f"First command: {first_command.commandstring}")

first_command_words = copy(first_command.words)
if mapped_execarg:
    first_command_words[0] = mapped_execarg

argtypes_initial, results_initial = interface.get_argtypes_and_results(
    interface_file, interface_py_file, interface_argindex, first_command_words, first_command.words[0]   
)

if interface_py_file is None:
    msg(2, "Try to obtain argtypes from rules")

    overrule_ext, overrule_no_ext = False, False
    if args.g1:
        msg(1, "disable argtyping guess rule 1")
        overrule_ext = True
    if args.g2:
        msg(1, "disable argtyping guess rule 2")
        overrule_no_ext = True

    try:
        argtypes_guess = guess_arguments(
            first_command_words, overrule_ext=overrule_ext, overrule_no_ext=overrule_no_ext
        )
    except ValueError as exc:
        msg(-1, *exc.args)
        sys.exit(1)
    if argtypes_initial is None:
        argtypes_initial = argtypes_guess
    else:
        argtypes_temp = argtypes_initial
        argtypes_initial = {}
        argtypes_initial.update(argtypes_guess)
        argtypes_initial.update(argtypes_temp)

argtypesstr = json.dumps(argtypes_initial, sort_keys=False, indent=2)
msg(1, "initial argtypes dict:\n" + argtypesstr)

workdir = os.getcwd()
if args.workdir is not None:
    workdir = os.path.expanduser(args.workdir)
    msg(1, "set argtyping working directory to: {}".format(workdir))

file_mapping_mode = args.file_mapping_mode
if file_mapping_mode is None:
    file_mapping_mode = "literal"
    
try:
    argtypes = get_file_mapping(
        argtypes_initial, mapping_mode=file_mapping_mode, working_directory=workdir
    )
except ValueError as exc:
    msg(-1, *exc.args)
    sys.exit(1)

argtypesstr = json.dumps(argtypes, sort_keys=False, indent=2)
msg(1, "argtypes dict:\n" + argtypesstr)

mapped_first_command = argtypes["@order"]

################################################################
assert len(mapped_first_command) == len(first_command.words)
word_substitutions = {}
for wordnr, word in enumerate(mapped_first_command):
    old_word = first_command.words[wordnr]
    if word != old_word:
        node = first_command.wordnodes[wordnr]
        word_substitutions[node] = word

################################################################
interface_data = {}
make_executables = []

def update_interface_data(new_interface_data, first):
    changed = False
    for k,v in new_interface_data.items():
        if k in ("argtypes", "files", "directories", "shim"):
            continue
        curr = interface_data.get(k)
        if k == "results":
            if first:
                continue
            if isinstance(v, list):
                v = {kk:None for kk in v}
            if k not in interface_data:
                interface_data[k] = {}
            if v:
                interface_data[k].update(v)
                changed = True
            continue
        if k == "environment":
            if curr is None:
                interface_data[k] = v
                continue
            if curr == v:
                continue
            raise NotImplementedError("Multiple environments")
        
        changed = True
        if curr is None:
            interface_data[k] = v
        else:
            if isinstance(v, list) != isinstance(curr, list):
                interface_data[k] = v
            elif isinstance(curr, list):
                interface_data[k] += v
            else:
                interface_data[k] = v
    if changed:
        msg(3, f"Updated interface data: {json.dumps(interface_data, indent=2)}")

for commandnr, command in enumerate(commands):
    if commandnr > 0:
        msg(1, f"Command #{commandnr+1}: {command.commandstring}")
        interface_argindex, interface_file, interface_py_file, mapped_execarg = interface.locate_files(command.words)        
    if interface_file is not None or interface_py_file is not None:
        if commandnr > 0:
            msg(1, f"Interface files found for command #{commandnr+1}: {command.commandstring}")
    
    shim = None
    if interface_file is not None:
        msg(2, f"loading {interface_file}")
        new_interface_data = interface.load(interface_file.as_posix())
        msg(3, f"{interface_file} content: {json.dumps(new_interface_data, indent=2)}")
        new_shim = new_interface_data.get("shim")
        if new_shim is not None:
            shim = new_shim
        update_interface_data(new_interface_data, first=(commandnr==0))

    if commandnr == 0:
        interface_py_file = None
    if interface_py_file is not None:
        if commandnr == 0:
            command_words = mapped_first_command
        else:
            command_words = copy(command.words)
            if mapped_execarg:
                command_words[0] = mapped_execarg

        arguments = command_words[interface_argindex+1:]
        new_interface_data = interface.interface_from_py_file(interface_py_file, arguments)
        msg(3, f"{interface_py_file} result: {json.dumps(new_interface_data, indent=2)}")
        new_shim = new_interface_data.get("shim")
        if new_shim is not None:
            shim = new_shim
        update_interface_data(new_interface_data, first=(commandnr==0))

        if shim:
            wordnode = command.wordnodes[interface_argindex]
            word_substitutions[wordnode] = shim
    
    if mapped_execarg and not shim:
        wordnode = command.wordnodes[0]
        word = command.words[0]
        word = word_substitutions.get(wordnode, word)
        make_executables.append(word)

################################################################

redirection = None
redirection_node = parsing.get_redirection(commands[-1])
if redirection_node is not None:
    redirection = redirection_node.word
    redirection = os.path.expanduser(redirection)
    if workdir is not None and redirection.startswith(workdir):
        redirection = redirection[len(workdir):]
    if redirection != redirection_node.word:
        word_substitutions[redirection_node] = redirection

if results_initial is None:
    results_initial = {}
else:
    msg(2, f"Initial result targets from first command: {results_initial}")    
    decal = os.path.relpath(os.getcwd(), workdir)
    if decal != ".":
        results_initial2 = results_initial.copy()
        for k, v in results_initial.items():
            if v is None:
                kk = os.path.join(decal, k)
                results_initial2.pop(k)
                results_initial2[kk] = v
        results_initial = results_initial2

results = results_initial.copy()
results.update(interface_data.get("results", {}))
if not results and not redirection:
    capture_stdout = True
    result_targets = None
else:
    capture_stdout = False
    result_targets = {}
    if results:
        result_targets.update(results)
    if redirection:
        result_targets[redirection] = None
    msg(1, f"Identify result targets: {result_targets}")

################################################################
docker_image = interface_data.get("environment", {}).get("docker_image")

env = {}

if docker_image is not None:
    msg(1, f'Set Docker image to "{docker_image}"')    
    env["docker"] = {"name": docker_image}

seamless.config.delegate()

# TODO: max_files, max_datasize option
paths = set()
directories = set()
mapping = {}
for argname, arg in argtypes.items():
    if isinstance(arg, dict):
        argtype = arg.get("type")
        path = arg.get("mapping", argname)
    else:
        argtype = arg
        path = argname
    if argtype in ("file", "directory"):
        paths.add(path)
        if argtype == "directory":
            directories.add(path)
        mapping[argname] = path

file_checksum_dict = files_to_checksums(
    paths, max_datasize=None, max_files=None, directories=directories
)
checksum_dict = {}
checksum_dict = {k: file_checksum_dict[v] for k,v in mapping.items()}
checksum_dictstr = json.dumps(checksum_dict, sort_keys=True, indent=2)
msg(2, "file/directory checksum dict:\n" + checksum_dictstr)
if len(directories):
    msg(2, "directories:", directories)

################################################################

for node in sorted(word_substitutions.keys(), key=lambda node: -node.pos[0]):
    word = word_substitutions[node]
    commandstring = commandstring[:node.pos[0]] + word + commandstring[node.pos[1]:]

msg(1, "bash command:\n", commandstring, "\n")

################################################################

command = commandstring.split()
variables = None
if len(commands) == 1:
    canonical = interface_data.get("canonical", [])
    for canon in canonical:
        vars = canon.get("variables", [])
        varnames = [var["name"] for var in vars]
        vartypes = {var["name"]:var["celltype"] for var in vars}
        canon_cmd = canon["command"].split()
        if len(canon_cmd) != len(command):
            continue
        for w1, w2 in zip(canon_cmd, command):
            if w1 == w2:
                continue
            if w1[0] == "$" and w1[1:] in varnames:
                continue
            break
        else:
            msg(1, f"Found canonical command match:\n  {canon['command']}")
            variables = {}
            for w1, w2 in zip(canon_cmd, command):
                if w1[0] == "$" and w1[1:] in varnames:
                    varname = w1[1:]
                    variables[varname] = (w2, vartypes[varname])
            msg(2, f"Variables:\n  {variables}")
            commandstring = canon["command"]

################################################################

#def write_pending(filename, )

transformation_checksum, transformation_dict = prepare_bash_transformation(
    commandstring, checksum_dict, directories=directories, 
    make_executables=make_executables,
    capture_stdout=capture_stdout, result_targets=result_targets,
    environment=env, variables=variables
)
if make_executables:
    msg(2, "The following files will be made executable inside the transformer\n:\n  {}".format('\n  '.join(make_executables)))
else:
    msg(3, "No files will be made executable inside the transformer")

from seamless import SeamlessTransformationError
try:
    result_checksum = run_transformation(transformation_dict, undo=args.undo)
except SeamlessTransformationError as exc:
    traceback.print_exc(limit=0)
    exit(1)

if not args.undo:
    msg(1, "Transformation finished")
    msg(2, "Result checksum: {}".format(result_checksum))

def write_result(filename, file_checksum):
    if workdir is not None and not os.path.dirname(filename):
        # File was created in the server workdir: grandfather it in
        pass
    else:
        filename = os.path.join(workdir, filename)
    file_checksum = Checksum(file_checksum)
    if file_checksum.value is None:
        return
    try:
        with open(filename + ".CHECKSUM", "w") as f:
            f.write(file_checksum.hex()+"\n")
    except Exception:
        msg(0, f"Cannot write checksum to result file '{filename}.CHECKSUM'")
        return
    try:
        file_buffer = buffer_cache.get_buffer(file_checksum.bytes())
        if file_buffer is None:
            raise CacheMissError(file_checksum)
    except CacheMissError:
        msg(0, f"Cannot obtain contents of result file '{filename}', CacheMissError")
    try:
        with open(filename, "wb") as f:
            f.write(file_buffer)
    except Exception:
        msg(0, f"Cannot write to result file '{filename}'")
        return

# kludge to hide spurious "buffers undestroyed" warnings
import logging
logger = logging.getLogger("seamless")
logger.setLevel(logging.ERROR)


if args.undo:
    if result_checksum is None:
        exit(1)
    else:
        msg(1, "Transformation undone")
        exit(0)

else:
    try:
        result_buffer = buffer_cache.get_buffer(result_checksum.bytes())
        if result_buffer is None:
            raise CacheMissError(result_checksum)
    except CacheMissError:
        traceback.print_exc(limit=1)
        exit(1)

    if (not capture_stdout) and len(result_targets) > 1:
        result_checksum_dict = deserialize(result_buffer, result_checksum.bytes(), "plain", copy=False)
        filenames = []
        for k, v in result_targets.items():
            filename = v if v is not None else k
            filenames.append(filename)      
        for filename in result_checksum_dict:
            assert filename in filenames, (filename, result_targets)
        for k, v in result_targets.items():
            filename = v if v is not None else k
            try:
                file_checksum = result_checksum_dict[k]
            except KeyError:
                msg(0, "No result for '{filename}' was returned")
                continue
            write_result(filename, file_checksum)
    elif result_targets and len(result_targets) == 1:
        k, v = list(result_targets.items())[0]
        filename = v if v is not None else k
        write_result(filename, result_checksum)
    elif capture_stdout:
        try:
            result = result_buffer.decode()
        except UnicodeDecodeError:
            result = result_buffer

        print(result)

