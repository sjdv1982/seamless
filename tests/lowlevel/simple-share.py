import seamless
from seamless.core import macro_mode_on
from seamless.core import context, cell, transformer, unilink, macro
from seamless.shareserver import shareserver
from seamless.core.share import sharemanager
from functools import partial

import websockets
import requests
import asyncio
import json
import time

shareserver_started = shareserver.start()

def define_ctx():
    with macro_mode_on():
        ctx = context(toplevel=True)
        ctx.cell1 = cell().set(1)
        ctx.cell2 = cell().set(2)
    ctx.compute()
    with macro_mode_on():
        ctx.result = cell()
        ctx.tf = transformer({
            "a": "input",
            "b": "input",
            "c": "output"
        })
        ctx.cell1_unilink = unilink(ctx.cell1)
        ctx.cell1_unilink.connect(ctx.tf.a)
        ctx.cell2.connect(ctx.tf.b)
        ctx.code = cell("transformer").set("c = a + b")
        ctx.code.connect(ctx.tf.code)
        ctx.result_unilink = unilink(ctx.result)
        ctx.tf.c.connect(ctx.result_unilink)
    return ctx

ctx = define_ctx()
ctx.compute()
name = sharemanager.new_namespace(ctx._get_manager(), True, name="ctx")
ctx.compute()
print("OK1", name)

print(ctx.cell1.value, ctx.cell2.value)
ctx.cell1.share(readonly=False)
ctx.cell2.share(readonly=False)
ctx.compute()

async def echo(uri):
    async with websockets.connect(uri) as websocket:
        async for message in websocket:
            print("WS ECHO", message)

ws = echo('ws://localhost:5138/ctx')
asyncio.ensure_future(ws)

loop = asyncio.get_event_loop()
loop.run_until_complete(shareserver_started)
loop.run_until_complete(asyncio.sleep(0.1))

def thread(func, *args, **kwargs):
    from threading import Thread
    from queue import Queue
    def func2(func, q, args, kwargs):
        result = func(*args, **kwargs)
        q.put(result)
    q = Queue()
    t = Thread(target=func2, args=(func, q, args, kwargs))
    t.start()
    while t.is_alive():
        t.join(0.05)
        loop.run_until_complete(asyncio.sleep(0.01))
    return q.get()

r = thread(requests.get, 'http://localhost:5813/ctx/cell1')
print(r.json())

r = thread(
    requests.put, 'http://localhost:5813/ctx/cell1',
    data=json.dumps({"buffer": "20\n"})
)

r = thread(requests.get, 'http://localhost:5813/ctx/cell1')
print(r.json())

ctx.cell1.set(99)
loop.run_until_complete(asyncio.sleep(0.1))

r = thread(requests.get, 'http://localhost:5813/ctx/cell1')
print(r.json())

ctx._get_manager().destroy()

ctx = context(toplevel=True)
name = sharemanager.new_namespace(ctx._get_manager(), True, name="ctx")
print("OK2", name)
assert name == "ctx"

ws = echo('ws://localhost:5138/ctx')
asyncio.ensure_future(ws)

def macro_code(ctx, param_a):
    ctx.a = cell().set(param_a + 1000)
    ctx.a.share()
    ctx.a0 = cell().set(999)
    ctx.a0.share()

def define_ctx2():
    ctx.macro = macro({"param_a": "int"})
    ctx.macro.code.cell().set(macro_code)
    ctx.param_a = cell().set(42)
    ctx.param_a.share(readonly=False)
    ctx.param_a.connect(ctx.macro.param_a)

define_ctx2()
import asyncio; asyncio.get_event_loop().run_until_complete(asyncio.ensure_future(asyncio.sleep(1)))
print(r.text)
print(ctx.param_a.value)
print(ctx.macro.ctx.a.value)

r = thread(requests.get, 'http://localhost:5813/ctx/macro/ctx/a')
print(r.json())

print("OK3")

r = thread(
    requests.put, 'http://localhost:5813/ctx/param_a',
    data=json.dumps({"buffer": "43\n"})
)
print(r.json())
asyncio.get_event_loop().run_until_complete(asyncio.sleep(0.5))  # to get the request processed
print("OK3a")
ctx.compute()

print(ctx.param_a.value)
print("OK3b")

ctx.compute()

print(ctx.param_a.value)
print(ctx.macro.ctx.a.value)
