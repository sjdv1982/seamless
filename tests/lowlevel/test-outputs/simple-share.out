Opened the seamless share update server at port 5138
Opened the seamless REST server at port 5813
OK1 ctx
1 2
WS ECHO ["Seamless share update server", "0.02"]
WS ECHO ["sharelist", ["cell1", "cell2", "self"]]
WS ECHO ["binary", ["cell1", "cell2"]]
1
WS ECHO ["update", ["cell1", "8e843baef228089dc379d4c3b6e28c1bb5d44eee257f1206b5dfee44ef6b05ad", 2]]
20
WS ECHO ["update", ["cell1", "bfb3843fc9199b316e2567251596f33f70784b6003b88d198b82be231352ddaa", 3]]
99
OK2 ctx
WS ECHO ["Seamless share update server", "0.02"]
WS ECHO ["sharelist", ["self"]]
WS ECHO ["binary", []]
WS ECHO ["sharelist", ["macro/ctx/a", "macro/ctx/a0", "param_a", "self"]]
WS ECHO ["binary", ["macro/ctx/a", "macro/ctx/a0", "param_a"]]
WS ECHO ["update", ["macro/ctx/a0", "a4654c74202cb987505719dc98609c3c46582d384f45f54659736280c66223f4", 1]]
WS ECHO ["update", ["param_a", "fa2fe6c9c0556871073be9a00d6d29bd3b9b6dd560587ee6e8c163755bf669d3", 1]]
WS ECHO ["update", ["macro/ctx/a", "0d5d541db3b11b331a8752b7e3c743ce43f0df6143589e4298f51027011a5937", 1]]
99

42
1042
1042
OK3
WS ECHO ["update", ["param_a", "5d5a3c73da8e376d31a96613e450d0cad333ebcf496acbc0eb9fa6c1ab8c42fa", 2]]
2
WS ECHO ["update", ["macro/ctx/a", "108cbc2435f5f35785d868b38892c9df7420159806175b336eaa62957d6c71ec", 2]]
OK3a
43
OK3b
43
1043
